src/overlays/actors/ovl_Obj_Dora/z_obj_dora.c

```diff
@@ -98,34 +152,188 @@ static DamageTable D_80B61488 = {
+
+void ObjDora_Init(Actor* thisx, GlobalContext* globalCtx) {
+    ObjDora* this = THIS;
+    s32 i, j;
+    Vec3f vtx[3];
+    s32 buf = 0, buff2 = 0;
+
+    Actor_SetScale(&this->actor, 0.1f);
+    ActorShape_Init(&this->actor.shape, 0.0f, &ActorShadow_DrawCircle, 36.0f);
+
+    this->gongRotation.x = 0;
+    this->gongRotation.z = 0;
+    this->gongAngle.x = 0;
+    this->gongAngle.z = 0;
+    this->lastGongHitType = DORA_HIT_NONE;
+    this->rupeeDropTimer = 0;
+    this->collisionCooldownTimer = 0;
+    this->unk3AA = 0;
+    this->gongForce.x = 0.0f;
+    this->gongForce.y = 0.0f;
+    this->gongForce.z = 0.0f;
+
+    Collider_InitTris(globalCtx, &this->colliderTris);
+    Collider_SetTris(globalCtx, &this->colliderTris, &this->actor, &sTrisInit, this->colliderTrisElements);
+    CollisionCheck_SetInfo2(&this->actor.colChkInfo, &sDamageTable, &sColChkInfoInit);
+
+    Matrix_SetStateRotationAndTranslation(this->actor.world.pos.x, this->actor.world.pos.y, this->actor.world.pos.z,
+                                          &this->actor.shape.rot);
+    Matrix_Scale(this->actor.scale.x, this->actor.scale.y, this->actor.scale.z, MTXMODE_APPLY);
+
+    for (i = 0; i < ARRAY_COUNT(this->colliderTrisElements); i++) {
+        for (j = 0; j < ARRAY_COUNT(vtx); j++) {
+            Matrix_MultiplyVector3fByState(&sTrisElementsInit[i].dim.vtx[j], &vtx[j]);
+        }
+        Collider_SetTrisVertices(&this->colliderTris, i, &vtx[0], &vtx[1], &vtx[2]);
+    }
+    ObjDora_SetupWait(this);
+}
+
+void ObjDora_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    ObjDora* this = THIS;
+
+    Collider_DestroyTris(globalCtx, &this->colliderTris);
+}
+
+void ObjDora_SetupWait(ObjDora* this) {
+    this->actionFunc = ObjDora_Wait;
+}
+
+void ObjDora_Wait(ObjDora* this, GlobalContext* globalCtx) {
+}
+
+void ObjDora_SetupMoveGong(ObjDora* this) {
+    if (this->lastGongHitType == DORA_HIT_LIGHT) {
+        this->gongForce.x = 2.0f;
+        this->gongForce.y = 5461.0f;
+        this->gongForce.z = 1820.0f;
+    } else {
+        this->gongForce.x = 4.0f;
+        this->gongForce.y = 12743.0f;
+        this->gongForce.z = 5461.0f;
+    }
+    this->gongAngle.x = 0;
+    this->gongAngle.z = 0;
+    this->actionFunc = ObjDora_MoveGong;
+}
+
+void ObjDora_MoveGong(ObjDora* this, GlobalContext* globalCtx) {
+
+    if ((this->gongForce.y < 182.0f) && (this->gongForce.z < 182.0f)) {
+        this->lastGongHitType = DORA_HIT_LIGHT;
+        ObjDora_SetupWait(this);
+    }
+
+    Math_SmoothStepToF(&(this->gongForce.x), 0, 0.2f, 0.2f, 0.1f);
+    Math_SmoothStepToF(&(this->gongForce.y), 0, 0.5f, 54.0f, 18.0f);
+    Math_SmoothStepToF(&(this->gongForce.z), 0, 0.5f, 54.0f, 18.0f);
+    this->gongAngle.x += 0x1555;
+    this->gongAngle.z += 0x238E;
+    this->gongRotation.x = Math_SinS(this->gongAngle.x) * this->gongForce.y;
+    this->gongRotation.z = Math_SinS(this->gongAngle.z) * this->gongForce.z;
+}
+
+bool ObjDora_IsHalfHourly(u16 time) {
```

Comment:
It's not clear to me what this function is doing, so I think it either needs better understanding of the floats, or some sort of explanatory comments.

Reply:
If you hit the gong at the the top/bottom of the hour the gong gives you a blue rupee. (IE deku link spin attack at 3PM day 1 or 3:30PM)

So this is some sort of time math, since its unlikely the player will hit at the exact frame of 3PM or half hour, they gave it a range by the looks of it.

Reply:
Also, we do not use `bool` (we can't: lots of different types are used as boolean values, basically every arithmetic type is used as a boolean somewhere in the codebase). Use s32 instead (but keep true/false).

Reply:
We do have a macro to convert float times to minutes... hopefully it gives sensible numbers here.

Reply:
I wasn't sure the best way to word the function, but yeah if the gong is hit on the 30 or hour mark a blue rupee drops. If I update with CLOCK_TIME_F think that's good? I was gonna change it to bihourly but that makes it a bit ambiguous.